onnx/large_helper.py

```diff
@@ -0,0 +1,324 @@
+            onnx.checker.check_model(self.model_proto)
+
+    @property
+    def model_proto(self) -> onnx.ModelProto:
+        if self.model_proto_ is None:
+            raise RuntimeError("ModelContainer is empty.")
+        return self.model_proto_
+
+    @model_proto.setter
+    def model_proto(self, model_proto: onnx.ModelProto):
+        self.model_proto_ = model_proto
+        self.graphs_ = list(self.enumerate_graph_protos())
+
+    def enumerate_graph_protos(self) -> Iterable[onnx.GraphProto]:
+        """
+        Enumerates all GraphProtos in a model.
+        """
+        yield self.model_proto.graph
+        yield from _enumerate_subgraphs(self.model_proto.graph)
+
+    def set_large_initializers(self, large_initializers: dict[str, np.ndarray]):
+        """
+        Adds all large tensors (not stored in the model).
+        """
+        for k in large_initializers:
+            if not k.startswith("#"):
+                raise ValueError(
+                    f"The location {k!r} must start with '#' to be ignored by check model."
+                )
+        self.large_initializers = large_initializers
+
+    def check_large_initializers(self):
+        for tensor in ext_data._get_all_tensors(self.model_proto):
+            if not ext_data.uses_external_data(tensor):
+                continue
+            prop: onnx.StringStringEntryProto | None = None
+            for ext in tensor.external_data:  # type: ignore[assignment]
+                if ext.key == "location":  # type: ignore[attr-defined]
+                    prop = ext
+            if prop is None:
+                raise RuntimeError(
+                    f"No location found for tensor name {tensor.name!r}."
+                )
+            if prop.value not in self.large_initializers:
+                raise RuntimeError(
+                    f"Unable to find large tensor named {tensor.name!r} "
+                    f"with location {prop.value!r} in "
+                    f"{sorted(self.large_initializers)}."
+                )
+
+    def _save_external(
+        self, file_path: str, all_tensors_to_one_file: bool
+    ) -> onnx.ModelProto:
+        """
+        Save the large model into a main onnx file and one file
+        per tensor. Follows the same format as :func:`write_external_data_tensors
+        <onnx.external_data_helper.write_external_data_tensors>`.
+        The main model needs to be modified to update the file location,
+        the function returns this modified copy.
+
+        Arguments:
+            file_path: model file
+            all_tensors_to_one_file: all tensors in one file
+
+        Returns:
+            modified main model proto
+        """
+        _unique_names: set[str] = set()
+
+        def _clean_name(prefix: str, name: str) -> str:
+            for c in ":/\\;,!":
+                name = name.replace(c, "")
+            base_name = name
+            i = 0
+            while name in _unique_names:
+                i += 1
```

Comment:
Looks like this is O(#_unique_names) which grows with the number of names? Would it be possible to simplify it?